bin/standalone-installer-unix

```diff
@@ -0,0 +1,178 @@
+#!/bin/bash
+#
+# Bash program to download the latest Nextstrain CLI standalone installation
+# archive for Linux or macOS, extract it into the current user's app data
+# directory, and check if PATH includes the installation destination.
+#
+# It maintains rough parity with the PowerShell program for Windows,
+# standalone-installer-windows.
+#
+# Set DESTINATION to change the installation location.
+#
+# Set VERSION to change the version downloaded and installed, or pass the
+# desired version as the first argument to this program.
+#
+set -euo pipefail
+shopt -s failglob
+
+: "${NEXTSTRAIN_DOT_ORG:=https://nextstrain.org}"
+
+# Globals declared here to make them more obvious, but set by main() to avoid
+# source ordering requirements and delay execution until the full file is
+# parsed.
+declare KERNEL TARGET DESTINATION VERSION
+
+# Wrap everything in a function which we call at the end to avoid execution of
+# a partially-downloaded program.
+main() {
+    KERNEL="${KERNEL:-$(uname -s)}"
+    TARGET="${TARGET:-$(target-triple)}"
+    DESTINATION="${DESTINATION:-$(destination)}"
+
+    VERSION="${1:-${VERSION:-latest}}"
+
+    local archive archive_url tmp
+
+    archive="standalone-${TARGET}.tar.gz"
+    archive_url="${NEXTSTRAIN_DOT_ORG}/cli/download/${VERSION}/${archive}"
+
+    # Move into a temporary working dir
+    tmp="$(mktemp -d)"
+    if ! debug; then
+        trap "rm -rf $tmp" EXIT
+    fi
+    pushd "$tmp" >/dev/null
+    log "Temporary working directory: $tmp"
+
+    log "Downloading $archive_url"
+    curl "$archive_url" \
+        --fail --show-error --location --proto "=https" \
+        > "$archive"
+
+    log "Extracting $archive"
+    mkdir standalone
+    tar xz$(if-debug v)f "$archive" -C standalone
+
+    if [[ -d "$DESTINATION" ]]; then
+        log "Removing existing $DESTINATION"
+        rm -rf "$DESTINATION"
+    fi
+
+    log "Installing to $DESTINATION"
+    mkdir -p "$(dirname "$DESTINATION")"
+    mv standalone "$DESTINATION"
+
+    popd >/dev/null
+
+    if ! debug; then
+        log "Cleaning up"
+        rm -rf "$tmp"
+    fi
+
+    cat <<~~
+______________________________________________________________________________
+
+Nextstrain CLI ($("$DESTINATION"/nextstrain --version)) installed to $DESTINATION.
+
+~~
+
+    if [[ "$(type -p nextstrain 2>/dev/null)" != "$DESTINATION/nextstrain" ]]; then
+        cat <<~~
+To make the "nextstrain" command available without using its full path,
+please add the following to your shell initialization file (e.g. ~/.bashrc or
+~/.zshrc):
+
+    export PATH="$DESTINATION:\$PATH"
+
+and then logout and login again.  You only need to do this once.
```

Comment:
Re: [should we do more to update `PATH`?](https://github.com/nextstrain/cli/pull/217#issuecomment-1230851374)

Two things that I've used personally come to mind:

1. Homebrew outputs a command at the end of the installation script that will append the necessary init command to the shell initialization file ([src](https://github.com/Homebrew/install/blob/805136bcfcbc40b95e476639fb2407773ab26445/install.sh#L995-L1021)).
2. Conda has its own `conda init` command which then writes a conditional block to the shell initialization file ([src](https://github.com/conda/conda/blob/c031405723cbf934de1a0b0c645b21e4c6fe5a7f/conda/core/initialize.py#L433)).

I think it'd be good to simplify down to one script block that the user can copy/paste and run without replacing any text (similar to Homebrew's behavior, and kinda option 2 from the linked comment). It would:

1. Detect the shell(s)
2. Determine the correct shell initialization file(s)
3. Append `PATH` modification to shell initialization file(s)

Reply:
Thanks for your thoughts here! This seems reasonable to me.

So we're aiming for something like [option 2 in my comment](https://github.com/nextstrain/cli/pull/217#issuecomment-1230851374). Should we take on any of the "optional sophistications" I noted which could apply here?

- Distinguishing between default vs. current shell, particularly if they're not the same (e.g. we're running in bash but default shell is zsh)
- Instructions for installing/modifying init files for all detected _installed_ shells (i.e. not just the default or current one)
- Future-proofing by recommending something like `source /home/tom/.local/share/nextstrain/cli/standalone/bashrc` or `source <(/home/tom/.local/share/nextstrain/cli/standalone/nextstrain init-shell bash)` instead of direct `PATH` modification.

I'm inclined to do the future-proofing bit to simplify (hypothetical) future changes for Linux/macOS users as they wouldn't have to update their shell init again, but the downside is it complicates our Windows install, as currently we can just automatically make the PATH adjustments in a shell-agnostic way. I'm also suspicious that the future-proofing would be worth it, as I can't currently imagine what other bits of shell init/config we'd need to do besides `PATH` (but my imagination may be limited).

Thoughts on any of the above? In particular, I wonder about the default vs. current shell question.

Reply:
> default vs. current shell question

Looks like this is what conda does or is capable of doing ([src](https://github.com/conda/conda/blob/c031405723cbf934de1a0b0c645b21e4c6fe5a7f/conda/core/initialize.py#L436)). I think it's a good idea if not too difficult.

> Future-proofing by recommending something like `source /home/tom/.local/share/nextstrain/cli/standalone/bashrc`

Do you mean having the user run `source /home/tom/.local/share/nextstrain/cli/standalone/bashrc` which then modifies `.bashrc`, or putting the one line in `.bashrc` itself?

I think the one line in `.bashrc` itself would be good. To add another example, `rbenv` added just one line to my `.zshrc` which I think is nice ([src](https://github.com/rbenv/rbenv/blob/585ed84283f3308380b843391ee7b12706ecff87/libexec/rbenv-init#L77)):

```sh
eval "$(rbenv init - zsh)"
```

Reply:
Ah, so the functionality of the conda source you linked is because you can do `conda init bash zsh` for example and it'll go and modify the config for each named shell. You can also ask it to modify the config for all shells it supports (quite a few). I'm asking about something a bit different. We need to detect the shell (X) to emit instructions for when we print something like, "Add this line to your X shell config: {instructions for X}". Should we detect the _current_ shell or the _default_ shell? This is a subtle thing, but they're not the same and we need to make a choice. I'd lean towards default shell.

Re: future-proofing: I don't mean having the user source something to do the modification for them. I mean the difference between instructing the user to add a line like this to their shell config:

    export PATH="/home/tom/.local/share/nextstrain/cli/standalone:$PATH"

or a line like this instead:

    source /home/tom/.local/share/nextstrain/cli/standalone/etc/bashrc

or maybe like this:

    source <(/home/tom/.local/share/nextstrain/cli/standalone/nextstrain init-shell bash)

The potential future-proofing benefit of the last two is that the necessary shell init is automatically updated when Nextstrain CLI is upgraded. With the first, if changes are ever necessary, we'd need to instruct users again to manually make them.

Reply:
Answering one of my own questions:

```
source <(/home/tom/.local/share/nextstrain/cli/standalone/nextstrain init-shell bash)
```

is preferable because we can write more complex conditional/detection logic in Python and emit simpler POSIX shell syntax instead of having to write separate shell scripts for each shell flavor.

Reply:
> Should we detect the _current_ shell or the _default_ shell? This is a subtle thing, but they're not the same and we need to make a choice. I'd lean towards default shell.

Oh right, nevermind. The _current_ shell in this installer is always going to be Bash because that's what we run the installer under. So it makes no sense to consider it! ISTR realizing this a week or two ago but clearly I'd forgotten again.

Reply:
Pushed a81d314 addressing this.

Reply:
I see, default shell is good then! Seems like it'd be easy to apply to all installed shells if wanted in the future.